source/tools/glkernel-cli/KernelJsonImporter.cpp

```diff
@@ -0,0 +1,183 @@
+#include "KernelJsonImporter.h"
+
+#include <glkernel/Kernel.h>
+
+#include <cppassist/logging/logging.h>
+
+void forEachCell(cppexpose::VariantArray * depthArray, std::function<void(const cppexpose::Variant&)> lambda)
+{
+    if (!depthArray)
+    {
+        cppassist::error() << "Malformed kernel input.";
+    }
+
+    for (const auto& heightVariant : *depthArray)
+    {
+        auto heightArray = heightVariant.asArray();
+
+        if (!heightArray)
+        {
+            cppassist::error() << "Malformed kernel input.";
+        }
+
+        for (const auto& widthVariant : *heightArray)
+        {
+            auto widthArray = widthVariant.asArray();
+
+            if (!widthArray)
+            {
+                cppassist::error() << "Malformed kernel input.";
+            }
+
+            for (const auto& elementVariant : *widthArray)
+            {
+                lambda(elementVariant);
+            }
+        }
+    }
+}
+
+// TODO: this is duplicated in the generated code, remove this duplication
+float variantToFloat(const cppexpose::Variant & v)
+{
+    return v.value<float>();
+}
+
+glm::vec2 variantToVec2(const cppexpose::Variant & v)
+{
+    const auto arr = v.asArray();
+
+    const auto x = variantToFloat(arr->at(0));
+    const auto y = variantToFloat(arr->at(1));
+
+    return glm::vec2(x,y);
+}
+
+glm::vec3 variantToVec3(const cppexpose::Variant & v)
+{
+    const auto arr = v.asArray();
+
+    const auto x = variantToFloat(arr->at(0));
+    const auto y = variantToFloat(arr->at(1));
+    const auto z = variantToFloat(arr->at(2));
+
+    return glm::vec3(x,y,z);
+}
+
+glm::vec4 variantToVec4(const cppexpose::Variant & v)
+{
+    const auto arr = v.asArray();
+
+    const auto x = variantToFloat(arr->at(0));
+    const auto y = variantToFloat(arr->at(1));
+    const auto z = variantToFloat(arr->at(2));
+    const auto w = variantToFloat(arr->at(3));
+
+    return glm::vec4(x,y,z,w);
+}
+
+KernelJsonImporter::KernelJsonImporter(const std::string& inputFileName)
+{
+    glkernel::kernel1 kernel1;
+    glkernel::kernel2 kernel2;
+    glkernel::kernel3 kernel3;
+    glkernel::kernel4 kernel4;
+
+    cppexpose::Variant root;
+
+    if (!cppexpose::JSON::load(root, inputFileName))
+    {
+        cppassist::error() << "Input JSON file could not be loaded.";
```

Comment:
replace with assertion, or manually return (is this even possible from constructor?)